File: src/lib/tiposkripto/src/BaseCheck.ts
import type { TestTypeParams_any } from "./CoreTypes.js";
import type { ITestResourceConfiguration } from "./types.js";

/**
 * BaseCheck is the internal unified base class for all verification phases.
 * It covers BDD's Then, AAA's Assert, and TDT's Validate.
 * This class is not exposed to users - use BaseThen, BaseExpected, or BaseAssert instead.
 */
export abstract class BaseCheck<I extends TestTypeParams_any> {
  public name: string;
  checkCB: (storeState: I["iselection"]) => Promise<I["then"]>;
  error: boolean;
  artifacts: string[] = [];
  status: boolean | undefined;

  constructor(
    name: string,
    checkCB: (val: I["iselection"]) => Promise<I["then"]>,
  ) {
    this.name = name;
    this.checkCB = checkCB;
    this.error = false;
    this.artifacts = [];
  }

  addArtifact(path: string) {
    if (typeof path !== "string") {
      throw new Error(
        `[ARTIFACT ERROR] Expected string, got ${typeof path}: ${JSON.stringify(
          path,
        )}`,
      );
    }
    const normalizedPath = path.replace(/\\/g, "/");
    this.artifacts.push(normalizedPath);
  }

  toObj() {
    const obj = {
      name: this.name,
      error: this.error,
      artifacts: this.artifacts,
      status: this.status,
    };
    return obj;
  }

  abstract verifyCheck(
    store: I["istore"],
    checkCB: (s: I["iselection"]) => Promise<I["isubject"]>,
    testResourceConfiguration: ITestResourceConfiguration,
    artifactory?: any,
  ): Promise<I["iselection"]>;

  async test(
    store: I["istore"],
    testResourceConfiguration: any,
    filepath: string,
    artifactory?: any,
  ): Promise<I["then"] | undefined> {
    const addArtifact = this.addArtifact.bind(this);

    try {
      const x = await this.verifyCheck(
        store,
        async (s: I["iselection"]) => {
          try {
            if (typeof this.checkCB === "function") {
              const result = await this.checkCB(s);
              return result;
            } else {
              return this.checkCB;
            }
          } catch (e) {
            this.error = true;
            throw e;
          }
        },
        testResourceConfiguration,
        artifactory,
      );
      this.status = true;
      return x;
    } catch (e) {
      this.status = false;
      this.error = true;
      throw e;
    }
  }
}

export type IChecks<I extends TestTypeParams_any> = Record<
  string,
  BaseCheck<I>
>;

File: src/lib/kafe/src/main/java/kafe/BaseCheck.java
package kafe;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.function.Function;

/**
 * BaseCheck is the unified base class for all verification phases.
 * It covers BDD's Then, AAA's Assert, and TDT's Validate.
 */
public abstract class BaseCheck<R> {
    public String name;
    public Function<R, Object> checkCB;
    public boolean error;
    public List<String> artifacts;
    public Boolean status;
    
    public BaseCheck(String name, Function<R, Object> checkCB) {
        this.name = name;
        this.checkCB = checkCB;
        this.error = false;
        this.artifacts = new ArrayList<>();
        this.status = null;
    }
    
    public void addArtifact(String path) {
        if (!(path instanceof String)) {
            throw new RuntimeException(
                "[ARTIFACT ERROR] Expected string, got " + path.getClass().getName() + ": " + path
            );
        }
        String normalizedPath = path.replace('\\', '/');
        artifacts.add(normalizedPath);
    }
    
    public abstract Object verifyCheck(
        R store,
        Function<R, Object> checkCB,
        ITTestResourceConfiguration testResourceConfiguration
    );
    
    public Map<String, Object> toObj() {
        Map<String, Object> obj = new HashMap<>();
        obj.put("name", name);
        obj.put("error", error);
        obj.put("artifacts", artifacts);
        obj.put("status", status);
        return obj;
    }
    
    public Object test(R store, ITTestResourceConfiguration testResourceConfiguration, String filepath) throws Exception {
        try {
            Object result = verifyCheck(store, checkCB, testResourceConfiguration);
            status = true;
            return result;
        } catch (Exception e) {
            status = false;
            error = true;
            throw e;
        }
    }
}
